pkg/routeagent_driver/handlers/ovn/submarinerNonGWRouteHandler.go

```diff
@@ -0,0 +1,181 @@
+)
+
+type NonGatewayRouteHandler struct {
+	event.HandlerBase
+	mutex           sync.Mutex
+	smClient        submarinerClientset.Interface
+	localEndpoint   *submarinerv1.Endpoint
+	remoteEndpoints map[string]*submarinerv1.Endpoint
+	k8sClientSet    clientset.Interface
+	isGateway       bool
+	transitSwitchIP string
+}
+
+func NewNonGatewayRouteHandler(smClientSet submarinerClientset.Interface, k8sClientset *clientset.Clientset) *NonGatewayRouteHandler {
+	// We'll panic if env is nil, this is intentional
+	return &NonGatewayRouteHandler{
+		smClient:        smClientSet,
+		remoteEndpoints: map[string]*submarinerv1.Endpoint{},
+		k8sClientSet:    k8sClientset,
+	}
+}
+
+func (nonGatewayRouteHandler *NonGatewayRouteHandler) Init() error {
+	logger.Infof("Starting NonGatewayRouteHandler")
+
+	node, err := nodeutil.GetLocalNode(nonGatewayRouteHandler.k8sClientSet)
+	if err != nil {
+		return errors.Wrap(err, "error getting the g/w node")
+	}
+
+	annotations := node.GetAnnotations()
+
+	transitSwitchIP, ok := annotations["k8s.ovn.org/node-transit-switch-port-ifaddr"]
+	if !ok {
+		logger.Infof("No transit switch IP configured")
+		return nil
+	}
+
+	nonGatewayRouteHandler.transitSwitchIP = transitSwitchIP
+
+	return nil
+}
+
+func (nonGatewayRouteHandler *NonGatewayRouteHandler) GetName() string {
+	return "submariner-nongw-route-handler"
+}
+
+func (nonGatewayRouteHandler *NonGatewayRouteHandler) GetNetworkPlugins() []string {
+	// TODO enable when we switch to new implementation
+	// return []string{cni.OVNKubernetes}
+	return []string{}
+}
+
+func (nonGatewayRouteHandler *NonGatewayRouteHandler) LocalEndpointCreated(endpoint *submarinerv1.Endpoint) error {
+	nonGatewayRouteHandler.mutex.Lock()
+	defer nonGatewayRouteHandler.mutex.Unlock()
+
+	nonGatewayRouteHandler.localEndpoint = endpoint
+
+	return nil
+}
+
+func (nonGatewayRouteHandler *NonGatewayRouteHandler) RemoteEndpointCreated(endpoint *submarinerv1.Endpoint) error {
+	nonGatewayRouteHandler.mutex.Lock()
+	defer nonGatewayRouteHandler.mutex.Unlock()
+
+	if !nonGatewayRouteHandler.isGateway || nonGatewayRouteHandler.transitSwitchIP == "" {
+		return nil
+	}
+
+	retryErr := retry.RetryOnConflict(retry.DefaultRetry, func() error {
+		_, err := nonGatewayRouteHandler.smClient.SubmarinerV1().
+			NonGatewayRoutes(endpoint.Namespace).Create(context.TODO(),
+			nonGatewayRouteHandler.getNonGatewayRoute(endpoint), metav1.CreateOptions{})
+		if !apierrors.IsAlreadyExists(err) {
+			return errors.Wrapf(err, "error processing the remote endpoint create event for %q", endpoint.Name)
+		}
+		return nil
+	})
+
+	return errors.Wrapf(retryErr, "processing the remote endpoint event failed even after retry")
+}
+
+func (nonGatewayRouteHandler *NonGatewayRouteHandler) RemoteEndpointRemoved(endpoint *submarinerv1.Endpoint) error {
+	nonGatewayRouteHandler.mutex.Lock()
+	defer nonGatewayRouteHandler.mutex.Unlock()
+
+	delete(nonGatewayRouteHandler.remoteEndpoints, endpoint.Name)
```

Comment:
I have a generic Q about EP storing in nonGatewayRouteHandler.remoteEndpoints or other handlers, 
we store a pointer to EP in this array, I can see that the pointer is sent from controller/watcher to all handlers, how do we make sure that the EP object isn't deleted/freed while we store a pointer to this object?

Reply:
Shouldn't that be taken care by golang memory management.

Reply:
Nice, 10x